drivers: pwm: stm32: add support for pinmux

gmarull opened the pull request:
Add support for DT based pinmux configurations on the PWM STM32 driver. All STM32 boards using PWM have been ported and H743Zi tested successfully. I do not have any F1 board to test the F1 remaps, so testing on that side would be appreciated.

NOTE: I have also done some minor driver cleanup (9bd0e6f), may be removed from this PR if needed.

gmarull committed:
drivers: pwm: stm32: remove data and config helpers

Now that Zephyr names have been made short and nice the need of these
helpers is less justified.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
drivers: pwm: stm32: add support for pinmux

Add support for DT based pinmux configurations.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_h743zi: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on Nucleo H743Zi
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f401re: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F401RE
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_l496zg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-L496ZG
board.

NOTE: pwm4 node has been deleted as it was not actually used nor
documented.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_h745zi_q_m7: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-H745ZI-Q
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_g431rb: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-G431RB
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_g474re: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-G474RE
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32f411e_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32F411E
discovery board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

lochej reviewed (commented):
Tested with blinky_pwm on nucleo_h743zi, red LED blinking as expected 👍

lochej commented:
@gmarull One idea, though, we could "script find" every SOC-STM32 boards using PWM/timers in there pinmux.c to identify which one need upgrade to DT pinctrl definition, so we don't miss any of them. Do you want to try that ?

lochej commented:
> @gmarull One idea, though, we could "script find" every SOC-STM32 boards using PWM/timers in there pinmux.c to identify which one need upgrade to DT pinctrl definition, so we don't miss any of them. Do you want to try that ?

@erwango @galak @gmarull 
I actually came up with a **little script** that tells every instance of PWM device of each board are still declared using pinmux.c instead of the new dt pinctrl method.

```bash
#!/bin/bash

pinmuxes=`find boards/arm/ -name "pinmux.c"`

#use Perl regex here to match with grep
#match for specific STM32 pinmux header, only used
#with ST based boards
match_stm32_pinmux="pinmux/stm32/pinmux_stm32.h"
match_peripheral_mux="pwm\d+"

for pinmux in $pinmuxes; do

	is_stm32_pinmux=`cat $pinmux | grep -cP $match_stm32_pinmux`
	if [ $is_stm32_pinmux -eq 0 ]
	then
		continue
	fi

	periph_mux_count=`cat $pinmux | grep -cP $match_peripheral_mux`
	periph_matches=`cat $pinmux | grep -oP $match_peripheral_mux`
	if [ $periph_mux_count -ne 0 ]
	then
		board=`dirname $pinmux`
		echo $board : $periph_matches
	fi
done
```

For now here is the list of matched board which still declare PWM pins using **pinmux.c** (The script execution is based on this PR branch so that's why H743ZI and all the boards you provided are not listed):

```
boards/arm/nucleo_f302r8 : pwm2
boards/arm/stm32f746g_disco : pwm3
boards/arm/nucleo_f756zg : pwm1
boards/arm/nucleo_f334r8 : pwm1
boards/arm/black_f407zg_pro : pwm2
boards/arm/olimexino_stm32 : pwm1
boards/arm/stm32vl_disco : pwm1
boards/arm/96b_stm32_sensor_mez : pwm3 pwm4 pwm9
boards/arm/nucleo_l4r5zi : pwm2
boards/arm/nucleo_f767zi : pwm1
boards/arm/nucleo_f413zh : pwm2
boards/arm/waveshare_open103z : pwm1
boards/arm/stm32l496g_disco : pwm2
boards/arm/black_f407ve : pwm2
boards/arm/96b_aerocore2 : pwm4 pwm5
boards/arm/stm32_min_dev : pwm1
boards/arm/olimex_stm32_h103 : pwm1
boards/arm/nucleo_l452re : pwm2
boards/arm/nucleo_l476rg : pwm2
boards/arm/blackpill_f411ce : pwm4
boards/arm/disco_l475_iot1 : pwm2
boards/arm/nucleo_f746zg : pwm1
boards/arm/nucleo_g071rb : pwm3
boards/arm/nucleo_f103rb : pwm1
boards/arm/stm32f4_disco : pwm2
boards/arm/b_l4s5i_iot01a : pwm2
boards/arm/stm32f3_disco : pwm1
boards/arm/mikroe_mini_m4_for_stm32 : pwm3
boards/arm/steval_fcu001v1 : pwm2
boards/arm/nucleo_f429zi : pwm1
boards/arm/nucleo_f412zg : pwm2
boards/arm/nucleo_wb55rg : pwm2
boards/arm/nucleo_l432kc : pwm2
```

gmarull commented:
Thanks @lochej! I’ll port the remaining ones.

gmarull committed:
boards: arm: nucleo_f302r8: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F302R8
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f746zg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F746ZG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f756zg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F756ZG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f334r8: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F334R8
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
west: update hal_stm32 revision

Update hal_stm32 revision so that access to missing pinctrl definitions
are accessible.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f407ve: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F407VE
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f407zg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F407ZG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: olimexino_stm32: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on OLIMEXINO-STM32
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32vl_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32VL-DISCO
board.

NOTE: missing pwm1 entry has been added.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: 96b_stm32_sensor_mez: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on i96Boards STM32
Sensor Mezzanine board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_l4r5zi: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-L4R5ZI
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f767zi: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F767ZI
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f413zh: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F413ZH
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: waveshare_open103z: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on Waveshare Open103Z
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32l496g_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32L496G-DISCO
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: 96b_aerocore2: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on 96Boards Aerocore2
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32_min_dev: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32 Minimum
Development board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: olimex_stm32_h103: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on Olimex STM32 H103
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_l452re: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-L452RE and
NUCLEO-L452RE-P boards.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_l476rg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-L476RG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: blackpill_f411ce: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on Blackpill F411CE
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: disco_l475_iot1: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on
DISCO-L475VG-IOT01A board.

NOTE: pinctrl file pointed to a wrong variant l475r(c-e-g)tx, however,
the SoC in the board corresponds to l475v(c-e-g)tx.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_g071rb: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-G071RB
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f103rb: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F103RB
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32f4_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32F4DISCOVERY
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: b_l4s5i_iot01a: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on B-L4SI-IOT01A
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32f3_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STM32F3-DISCOVERY
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: mikroe_mini_m4: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on Mikroe mini M4 for
STM32 board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: steval_fcu001v1: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on STEVAL-FCU001V1
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f429zi: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F429ZI
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_f412zg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-F412ZG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_wb55rg: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-WB55RG
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: nucleo_l432kc: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on NUCLEO-L432KC
board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: arm: stm32f746g_disco: use DT for PWM pinmux configuration

Use the new DT facilities to configure PWM pinmux on
STM32F746G-DISCOVERY board.

NOTE: DT entry for timers1/pwm1 has been replaced by timers3/pwm3 as it
was the pwm referenced in the pinmux file.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

lochej commented on boards/arm/nucleo_l452re/nucleo_l452re.dts:
```diff
@@ -7,6 +7,7 @@
 
 /dts-v1/;
 #include "nucleo_l452re_common.dtsi"
+#include <st/l4/stm32l452r(c-e)tx-pinctrl.dtsi>
 #include "arduino_r3_connector.dtsi"
```
I think it is better to do this inclusion is the nucleo_l452re_common.dtsi file as you will reference pinctrl in this file instead of variant (P and non-P). And in fact the common dtsi doesn't include the pinctrl definitions... 🙃

lochej replied:
Just noticed the 2 inclusions are different... my bad, because revision P has different pinctrl definitions ! I remove my comments !

lochej commented on boards/arm/nucleo_l452re/nucleo_l452re_p.dts:
```diff
@@ -7,6 +7,7 @@
 
 /dts-v1/;
 #include "nucleo_l452re_common.dtsi"
+#include <st/l4/stm32l452retxp-pinctrl.dtsi>
 #include "arduino_r3_connector.dtsi"
```
Same here, might be best placed in nucleo_l452re_common.dtsi

lochej replied:
Just noticed the 2 inclusions are different... my bad, because revision P has different pinctrl definitions ! I remove my comments !

lochej reviewed (commented):
Here is a single comment for the specific nucleo_l452re 👍 
BTW good job for porting other boards too !

lochej reviewed (approved):
LGTM ! Thanks for porting other boards

lochej commented:
@gmarull We might need to split this PR into multiple to add less boards at one time. Looks like buildkite timed out because it tried to run tests on each boards. Or else is there anither solution ?

erwango commented:
@gmarull, to get this PR passing BK CI, please add a temp commit on top of it that will tweak BK pipeline:
In `.buildkite/pipeline.yml`:
```
-    parallelism: 20
+    parallelism: 50
```
This patch will have to be removed just before merge.

erwango reviewed (approved):
Thanks for this comprehensive PR!
Tested ok on the few boards I had on my desk (nucleo_f429zi, disco_l476_iot1, nucleo_wb55rg, nucleo_f103rb, nucleo_f746rg).

gmarull commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -276,6 +281,127 @@ static int pwm_stm32_init(const struct device *dev)
+			if (remap == REMAP_FULL) {
+				LL_GPIO_AF_EnableRemap_TIM4();
+			}
+			break;
```
Regarding Timer 5, the only available remap function is `LL_GPIO_AF_EnableRemap_TIM5CH4`. As stated in the summary, I'm not sure how to handle a channel dependent remap. Timer 8 has no remap options, at least on LL API.

erwango replied:
Let's keep it as is for now. If need arises, the usage will become clearer.

ABOSTM replied:
Oups sorry, I forgot the point you mentioned in the summary, after the review of the 44 commits.
Ok for TIM8.
About TIM5 remap, even if the channel pin cannot really be remapped, it can still be disabled by setting bitfield TIM5CH4_IREMAP
![image](https://user-images.githubusercontent.com/50730894/95062097-c74a6980-06fc-11eb-8ccc-8d87d6dfbeeb.png)
So I would say it is safer to call LL_GPIO_AF_DisableRemap_TIM5CH4() to make sure remap TIM5_CH4 is connected to PA3.
```
#if DT_NODE_HAS_STATUS(DT_NODELABEL(timers5), okay)
		case DT_REG_ADDR(DT_NODELABEL(timers5)):
			LL_GPIO_AF_DisableRemap_TIM5CH4();
			break;
#endif`
```
But it seems that generated pinctrl files fot F1 doesn't contains anything about TIM5.

gmarull replied:
ST database does not contain anything for TIM5. As the remap is for internal routing, PWM will never need such remap anyway. Regarding the call to DisableRemap, should we call it when `NO_REMAP` is found? If yes, we need to update all other cases as it is not handled. UART driver will also need to be updated to handle this case.

ABOSTM replied:
> Regarding the call to DisableRemap, should we call it when `NO_REMAP` is found? If yes, we need to update all other cases as it is not handled. UART driver will also need to be updated to handle this case.

I think so. Otherwise we rely on hypothesis that nobody changes this (we remain on default value). And maybe this hypothesis is valid today, but maybe not on the future.

gmarull replied:
@ABOSTM fixed

erwango replied:
> UART driver will also need to be updated to handle this case.

I'm pushing a fix for this one

gmarull replied:
@ABOSTM can you review latest changeset?

ABOSTM commented on drivers/pwm/pwm_stm32.c:
```diff
@@ -286,6 +281,153 @@ static int pwm_stm32_init(const struct device *dev)
+			if (remap == REMAP_FULL) {
+				LL_GPIO_AF_EnableRemap_TIM13();
+			} else {
+				LL_GPIO_AF_DisableRemap_TIM12();
```
typo
```suggestion
				LL_GPIO_AF_DisableRemap_TIM13();
```

gmarull replied:
good catch, fixed

ABOSTM reviewed (approved):
LGTM

gmarull commented:
@galak This PR should be ready, 930a3b7 needs to be removed before merging.

galak commented:
dropped the buildkite commit and merging as it passes CI.
